torch/fx/experimental/symbolic_shapes.py

```diff
@@ -500,6 +514,22 @@ def free_unbacked_symbols_with_path(
             ):
                 r[s] = path
                 pending.remove(s)
+            # When an unbacked SymInt is perfectly divisible by an integer
+            # constant, we replace it with the integer constant to improve
+            # reasoning capabilities.  However, in synthetic examples, it is
+            # then possible that the factor never is explicitly allocated.
+            # Fortunately, we can compute it by division.
+            elif (
```

Comment:
I really dislike this single-casing.
I think a better way to do this would be to declare that we just allow to do unbacked symint replacements by fresh symints iff they are bijective, and we ask to provide an inverse.
We do this through a different API, and we ban replacements of unbacked symints by other symints in any other API (as you do here https://github.com/pytorch/pytorch/pull/124316).

Reply:
By doing this, we also avoid having to check explicitly for division, because we know that this is only set under the divisibility assumption.

Reply:
I'm not entirely sure exactly what you're proposing. In particular, at the end of the day, I need to write an inverse and then apply it when I'm looking for bindings. Are you suggesting that we just move this code as is into the point where we do the divisibility replacement, so it is extensible? I suppose we could eventually do it, but with only one weird case like this, I'd rather just keep it manual and refactor into some sort of open registration API later if we actually need it.

Or perhaps you are suggesting that we don't actually do the replacement, and then somehow later do the binding for u0?

If we're annoyed about the complexity here, I think my preference would be to think about how we could just get rid of the divisibility replacement trick entirely.

Reply:
>re you suggesting that we just move this code as is into the point where we do the divisibility replacement

Yes, this, but sure, let's do it at a later time.

Somewhat tangential to this, it'd be good to add a note discussing what needs to be touched to add an unbacked symint replacement with an unsupported formula, as this is so far from the location where you do the replacement.